### DISCUSSION TITLE
Unable to resolve dependency tree for desktop-viewer-react

### DISCUSSION THREAD

**Discussion by HughTheChapman** [iTwin/itwinjs-core#6268] [Q&A]
Hello,

I am trying to add the "desktop-viewer-react" package to an existing Electron desktop application to embed a viewer into it. I've been following the instructions from the repository: https://github.com/iTwin/viewer/tree/master/packages/modules/desktop-viewer-react. However, after I execute the command in the repository's installation instructions (npm install @itwin/desktop-viewer-react) I get an error.

![Terminal Error](https://github.com/iTwin/itwinjs-core/assets/133032675/f2c8971d-259d-4abf-9084-b123f00044d0)
 
What I've tried:

I've tried tinkering with the versions for many of the application's dependencies to attempt to get it working. I've used the example application that is created with "npx create-react-app@latest my-app --scripts-version @bentley/react-scripts --template @itwin/desktop-viewer" as a reference for which versions to change the dependencies to.

I also tried deleting the node_modules in that created example application and running "npm install" to test that my npm could put those packages together. And it could just fine.

Lastly, I've tried changing the node and npm version (using nvm) to what is recommended by this chart: https://www.itwinjs.org/learning/supportedplatforms/. That ended up being node v18.19.0 and npm v10.2.3. Didn't change the error any.

Any ideas on what might be causing the dependency conflict and/or how to fix it?

Thanks for the help,
Hugh

**Comment by aruniverse** [comment]
Hey Hugh!

Sorry to hear that. While we look to resolve the above error, have you tried using `npm install` with `--legacy-peer-deps` like the error message says?

**Reply by HughTheChapman** [reply]
I had tried --force but I don't think I had tried --legacy-peer-deps. I just tried it now, it did finish running the command. However, it then had a very similar error when running the application.

![Terminal Error 2](https://github.com/iTwin/itwinjs-core/assets/133032675/40dca904-7be3-45fb-9cc5-4ca102042c9e)

Since this error looked like it was complaining more about electron I tried changing it's version. But then the error looked much more similar to the one I had gotten originally. I messed around with the dependencies for a bit and didn't manage to get a configuration that would run.

I also tried running "npm run start --legacy-peer-deps" and it doesn't seem to be able to run that way.

![Terminal Error 3](https://github.com/iTwin/itwinjs-core/assets/133032675/c77c1975-853b-4b98-8192-dcdb87fe9a9a)

**Reply by aruniverse** [reply]
Can you try @itwin/desktop-viewer-react@4.2.3? I updated the peers yesterday.
Seems like you are pulling in two different versions of electron.

Could you create a github repo with your code so we can take a look at it?

**Reply by HughTheChapman** [reply]
The project is private so unfortunately I can't show the code. How would you go about embedding an iTwin Viewer in a template electron-vite application? I may be able to work it out from that.

**Reply by aruniverse** [reply]
Did the updated desktop-viewer-react pkg work?

**Reply by HughTheChapman** [reply]
Ah, forgot to check that. I tried it now and it does run 'npm install @itwin/desktop-viewer-react' just fine now. So that seems like progress! But, it seems to encounter several errors when starting the app. I tested it a few times and it only fails to run if any code that uses desktop-viewer-react runs. Otherwise it runs as normal.

There's quite a lot of errors. They all look to be module parse failures (so at least it is consistent). They all say "You may need an appropriate loader to handle this file type, currently no loaders are configured to process this file. See https://webpack.js.org/concepts#loaders". I looked at the link. I didn't manage to understand much of it; I'm not familiar with Webpack.

Here's the console log in a txt for reference: [output.txt](https://github.com/iTwin/itwinjs-core/files/13621506/output.txt)

Would this error simply be on my end?

**Reply by HughTheChapman** [reply]
I looked into Webpack over the weekend and I managed to fix all of the errors regarding the loaders by installing some loaders. After that I still had 2 errors left that occur when running the app. Errors for reference:  [output2.txt](https://github.com/iTwin/itwinjs-core/files/13638769/output2.txt)

I managed to get past them by adding this code snippet to webpack.config.js:

![CodeSnippet](https://github.com/iTwin/itwinjs-core/assets/133032675/607657a3-6683-4088-882b-86abab61e222)

So now the app is in a state where it can run and operate as normal. I want to test to make sure that the desktop-viewer-react package has functionality before calling this solved, and I'm having trouble with that. When I try to render viewer as a component nothing shows up. But at least it doesn't crash or anything!

This is the viewer component I use, I found it from the dekstop-viewer-react repository:

![ViewerComponent](https://github.com/iTwin/itwinjs-core/assets/133032675/047a5e65-e54b-4323-a36c-262ea3b9321f)

And this is how I attempt to render it (I've confirmed this works with other React components):

![RenderCode](https://github.com/iTwin/itwinjs-core/assets/133032675/14a93251-80f3-4738-a621-bbf96fb59249)

Could this mean the package isn't being used by the program correctly or is it more likely the code I'm using has a problem?
Thanks for the help so far!

**Reply by aruniverse** [reply]
You need to initialize the desktop viewer first using the `useDesktopViewerInitializer` hook before you can render the component

**Reply by HughTheChapman** [reply]
I can't find any documentation on that. Is there any you could point me to?

**Reply by aruniverse** [reply]
Docs are minimal, but you can find the implementation here: https://github.com/iTwin/viewer/blob/master/packages/modules/desktop-viewer-react/src/hooks/useDesktopViewerInitializer.tsx

I recommend reviewing how the desktop viewer template sets up the Viewer (take a look at the initializer hook usage and how the component gets rendered)

**Reply by HughTheChapman** [reply]
I tried replicating the implementation from the viewer template but the viewer still doesn't render. I checked the value of my "initialized" boolean and it returns false. As opposed to the template's which returns true. I assume this is why it won't render still. Do you know under what circumstances "initialized" returns false?

Note: I commented out the code in OnIModelAppInit since installing the packages required for that code brought up dependency issues with React. I put an alert() in OnIModelAppInit to see if it even runs OnIModelAppInit with "initialized" returning false and it doesn't seem to. So, I don't think that would be causing this particular problem. Though, I could be wrong.

Snippet of the code for reference:
![CodeSnippet2](https://github.com/iTwin/itwinjs-core/assets/133032675/277284f0-d272-4fce-b3f7-da0bcc1255c2)
